docker/src/main/java/clocker/docker/location/DockerLocation.java

```diff
@@ -245,13 +239,20 @@ public MachineLocation obtain(Map<?,?> flags) throws NoMachinesAvailableExceptio
         Entities.waitForServiceUp(dockerHost);
 
         // Obtain a new Docker container location, save and return it
-        if (LOG.isDebugEnabled()) {
+        try {
             LOG.debug("Obtain a new container from {} for {}", machine, entity);
+            Map<?,?> hostFlags = MutableMap.copyOf(flags);
+            DockerContainerLocation container = machine.obtain(hostFlags);
+            containers.put(machine, container.getId());
+            return container;
+        } finally {
+            // Release any placement strategy locks
+            for (DockerAwarePlacementStrategy strategy : Iterables.concat(strategies, entityStrategies)) {
+                if (strategy instanceof WithMutexes) {
+                    ((WithMutexes) strategy).releaseMutex(entity.getApplicationId());
```

Comment:
My gut feel (not fully thought through) was that if `strategy.filterLocations` fails to find a machine then we'd provision a new machine and then ask the strategy again which machine it wanted. The plus side of that would be that we don't need to keep holding the lock for a long time. The downsides are: 1) if we're not holding the lock, then then new machine could get grabbed by another app so our original obtain request might get starved; and 2) if the strategy keeps saying "no" then we'd keep getting more machines (but we'd guard against that by failing after a certain number of attempts, perhaps?).

I'm fine with leaving it as-is. But worth some more discussion.